### PR TITLE
New Native Fiat Plugin for credit cards

### PR DESCRIPTION
**Pull request by paullinator**
This PR enabled a new Native GUI Fiat plugin architecture. Full React Native UI (no more EdgeProvider webview) allowing a code-only plugin to launch various UI components such as an EnterAmount scene, wallet picker, List Modal, and a Webview. 

Currently implemented in this PR is a credit card plugin that uses two dummy providers and a fully functional Moonpay provider. Dummy providers are useful in testing best price hunting and user override of which provider to use.

Getting this merged will not affect behavior of app as it is not enabled yet via the buyPluginList.json file
Screen Shot 2022-06-19 at 16 03 28
4d6f-ba23-a6523e6fb9a8.png">
Screen Shot 2022-06-19 at 16 03 19

Screen Shot 2022-06-19 at 16 06 42

Screen Shot 2022-06-19 at 16 10 21

Screen Shot 2022-06-19 at 16 09 11

#### PR Requirements

If you have made **any** visual changes to the GUI. Make sure you have:
- [x] Tested on iOS device
- [x] Tested on Android device
- [x] Tested on small-screen device (iPod Touch)
- [x] Tested on large-screen device (tablet)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202470137026267

### CODE REVIEW

File: src/plugins/gui/creditCardPlugin.js [hunk 1]
```diff
@@ -161,11 +166,47 @@ export const creditCardPlugin: FiatPluginFactory = async (params: FiatPluginFact
...
+
+              // 2. Set the best quote to the one chosen by user (if any is chosen)
+              if (rowName != null && rowName !== bestQuote.pluginDisplayName) {
+                bestQuote = goodQuotes.find(quote => quote.pluginDisplayName === rowName)
+                if (bestQuote == null) return
+
+                // 3. Set the status text and powered by
+                const statusText = getRateFromQuote(bestQuote)
+                enterAmountMethods.setStatusText({ statusText })
+                enterAmountMethods.setPoweredBy({ poweredByText: bestQuote.pluginDisplayName, poweredByIcon: bestQuote.partnerIcon, poweredByOnClick })
```
Comment (swansontec) [inline src/plugins/gui/creditCardPlugin.js:216]: Oh, I get it - we want this to be interactive, so the user can change the powered-by text by tapping.

I wonder if we can clean this up to avoid the whole `getMethods` round-trip. Maybe the top-level `enterAmount` method returns an `updateProps` method, in addition to the results promise, so we have a single location to change anything we need, including future stuff we might add or existing things like labels.

File: src/plugins/gui/fiatProviderTypes.js [hunk 2]
```diff
@@ -61,7 +61,8 @@ export type FiatProvider = {
 export type FiatProviderFactoryParams = {
   io: {
     store: Object
-  }
+  },
+  apiKeys?: Object | string
```
Comment (swansontec) [inline src/plugins/gui/fiatProviderTypes.js:68]: I think this should be `unknown`, and then expect the plugins to use `asString` / `asObject({ something: asString })` to validate them.

**Review by swansontec — CHANGES_REQUESTED** [review]
Overall this looks super-nice! The overall direction is fantastic, but some of the individual back & forth API's could be streamlined in various ways.

File: src/plugins/gui/fiatPluginTypes.js [hunk 3]
```diff
@@ -9,7 +9,8 @@ export type FiatPluginStore = {
 }
 
 export type FiatPluginGetMethodsResponse = {
-  setStatusText: ({ statusText: string, options?: { textType?: 'warning' | 'error' } }) => void
+  setStatusText: ({ statusText: string, options?: { textType?: 'warning' | 'error' } }) => void,
+  setPoweredBy: ({ poweredByIcon: string, poweredByText: string, poweredByOnClick: () => void }) => void
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/plugins/gui/fiatPluginTypes.js]: Can't be a prop as the plugin is raw code and can't change props after the component is mounted. The labels don't ever change hence why they are props.

File: src/plugins/gui/fiatProviderTypes.js [hunk 4]
```diff
@@ -36,7 +51,7 @@ export type FiatProviderGetQuoteParams = {
 export type FiatProvider = {
   pluginId: string,
   getSupportedAssets: () => Promise<FiatProviderAssetMap>,
-  getQuote: (params: FiatProviderGetQuoteParams) => Promise<FiatProviderQuote | void>
+  getQuote: (params: FiatProviderGetQuoteParams) => Promise<FiatProviderQuote | FiatProviderQuoteError | void>
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/plugins/gui/fiatProviderTypes.js:61]: Sounds good. I like it. I'll try to rebase those changes but that might be a beast.

File: src/plugins/gui/fiatProviderTypes.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+export type FiatProvider = {
+  pluginId: string,
+  getSupportedAssets: () => Promise<FiatProviderAssetMap>,
+  getQuote: (params: FiatProviderGetQuoteParams) => Promise<FiatProviderQuote | void>
+}
+
+export type FiatProviderFactoryParams = {
+  io: {
+    store: Object
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/plugins/gui/fiatProviderTypes.js:66]: Ah. I didn't realize each key was a file. Yikes. Ok I'll mirror `EdgeDataStore` and just remove the storeId. Maybe `PluginDataStore` as a name

File: src/plugins/gui/fiatProviderTypes.js (hunk 2 above)
Comment (paullinator) [reply src/plugins/gui/fiatProviderTypes.js:68]: Ok. That's fair.

File: src/plugins/gui/fiatProviderTypes.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (paullinator) [reply src/plugins/gui/fiatProviderTypes.js:61]: Oh yeah, the `amount` is not a string because we would never use this as a precise amount that would need to be sent to an exchange. Using a js number makes for each comparison. at the plugin level.

Every other amount is a string because

1. crypto amounts might need full precision if we support Sell and an exchange is expecting perfect accuracy.
2. while fiat amounts never need that much accuracy, it's easier to be consistent with crypto amounts

[the remark this answers is not in the document]
Comment (paullinator) [reply src/plugins/gui/fiatProviderTypes.js:61]: The more I look into it, changing to using Error objects would require using `Promise.allSettled` on the plugin side since we were using `Promise.all` across all calls to `getQuote`. A few issues and inconveniences here:

1. Promise.allSettled isn't well supported on RN. That's ok, it's 4 line function.
2. Promise.allSettled kind of just takes error objects and puts them into a regular object in the returned array of Promise.allSettled. Sort of back to what we were doing.
3. The errors we get in the resulting array will be a mix of FiatQuote errors and other errors caused by a generic throw. So we couldn't really use `error instanceof FiatQuoteLimitError` to check if we got an error we care about so we'd have to check for instanceof every type of FiatQuote error possible. 
4. In the current implementation, the resulting array from getQuote calls has either `undefined` which were timeouts or generic throws, QuoteError objects with a known param we can check (`errorType`), and everything else we know is a good quote.
5. Declaration of the different errors is very verbose: ie

```
export class FiatProviderQuoteLimitError extends Error {
  type: 'overLimit' | 'underLimit'
  amount: number

  constructor(type: 'overLimit' | 'underLimit', amount: number) {
    super('Amount is outside of limits')
    this.type = type
    this.amount = amount
  }
}
```
vs
```
export type FiatProviderQuoteError = { errorType: 'overLimit' | 'underLimit', errorAmount: number }
```

So in summary, while I did give it an honest effort, I'm not sure what benefit we're getting by using Error objects.

File: src/types/GuiPluginTypes.js [hunk 6]
```diff
@@ -24,6 +24,9 @@ export type GuiPlugin = {
   // Also used for conversion tracking:
   storeId: string,
 
+  // Is a native plugin written in React Native
+  nativePlugin?: boolean,
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/types/GuiPluginTypes.js:28]: I'm a little confused here. Could you explain further?

File: src/plugins/gui/fiatProviderTypes.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (swansontec) [reply src/plugins/gui/fiatProviderTypes.js:61]: We did it this way in the swap scene, and it came out nicely. Network failures and other issues force you to handle “real” errors anyhow, and we don’t want misbehaving plugins to lock out working plugins, so while it was more work this way, the reliability of a single code-path was well worth it. All the utilities needed are already in the core.

[the remark this answers is not in the document]
Comment (paullinator) [reply src/plugins/gui/fiatProviderTypes.js:61]: Not sure I understand this. The way it’s coded, no misbehaving plug-ins could cause others to fail. 

Can you give an example of a “real” error that we would have to handle from a plug-in that we don’t do currently? Unless the real errors have an actual error object (and we can always add more) then we don’t have any way to handle them other than to ignore that one plug-in.

[the remark this answers is not in the document]
Comment (paullinator) [reply src/plugins/gui/fiatProviderTypes.js:61]: Also, error objects don't allow us to enforce what type of errors can be thrown. You can enforce return types but not thrown errors. We would require developers to comply with the standard.

File: src/types/GuiPluginTypes.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (paullinator) [reply src/types/GuiPluginTypes.js:28]: Ok. understood. will do

File: src/plugins/gui/fiatPluginTypes.js [hunk 7]
```diff
@@ -1,3 +1,50 @@
...
+
+export type FiatPluginStore = {
+  writeData: (data: { [key: string]: string }) => Promise<{ [key: string]: boolean }>,
+  readData: (keys: string[]) => Promise<{ [key: string]: string }>
+}
+
+export type FiatPluginEnterAmountParams = {
+  headerTitle: string,
+  label1: string,
+  label2: string,
```
[the remark this answers is not in the document]
Comment (swansontec) [reply src/plugins/gui/fiatPluginTypes.js:14]: Ok, then that makes sense.

File: src/plugins/gui/creditCardPlugin.js [hunk 8]
```diff
@@ -170,9 +172,11 @@ const ERROR_TEXT = {
   assetUnsupported: s.strings.fiat_plugin_asset_unsupported
 }
 
-export const getBestError = (errorQuotes: FiatProviderQuoteError[], currencyCode: string): string | void => {
+export const getBestError = (errorQuotes: FiatProviderError[], currencyCode: string): string | void => {
```
Comment (swansontec) [inline src/plugins/gui/creditCardPlugin.js:237]: The `errorQuotes` array will eventually need to be `mixed[]`, since we don't actually know what is being thrown. I think TypeScript will eventually complain if / when we upgrade to that, but for now Flow is happy so we can leave it. The overall architecture is correct, even if the types are loose.

**Review by swansontec — COMMENTED** [review]
OK, these fixups look good to me! @thehobbit85 can take over now and review the code itself. The basic architecture seems solid at this point.

File: src/plugins/gui/scenes/EnterAmountScene.js [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+      }
+    })
+  }
+  firstRun.current = false
+  let headerIcon = null
+  if (headerIconUri != null) {
+    headerIcon = <Image style={styles.icon} source={{ uri: headerIconUri }} />
+  }
+
+  const handleChangeText1 = useCallback(
```
Comment (paullinator) [inline src/plugins/gui/scenes/EnterAmountScene.js:62]: useHandler

File: src/components/themed/OutlinedTextInput.js [hunk 10]
```diff
@@ -294,7 +305,7 @@ export const OutlinedTextInput: Class<OutlinedTextInputRef> = forwardRef((props:
           autoFocus={autoFocus}
           multiline={multiline}
```
Comment (paullinator) [inline src/components/themed/OutlinedTextInput.js:306]: disable text input when showing spinner
